rt4914/KlinicCon-Backend
# Fixes #45: Creation of empty patient or doctor profile during sign up

1. application_controller.rb is updated with the new devise parameter sanitizer so that user_type is added as new one to the strong parameters list for only sign up
2. For doctor model, added optional true for all belongs_to relationships as without that there will be error while creating the empty doctor profile
3. For patient model, added optional true for address relationship so that there will be no error while creating the empty patient profile
4. Overriding the create method in registrations controller so that user can be saved with either empty patient or doctor profile, logic added to check the user type and accordingly blank profile is added
5. One public method create_profile is added under user model so that after user is created patient or doctor blank profile is created based on the user_type found in params hash


Resolves #45 

### Description

1. Fixes feature where during signup we will be creating empty patient or doctor profile based on user type i.e., type can be either patient or doctor

### Type of change


### How to test this PR?

1. Using sign up request body provide user_type as doctor as shown below

![CleanShot 2024-10-20 at 15 24 56](https://github.com/user-attachments/assets/db478424-c8ec-42cf-8899-437a2407ec11)

2. Send the request below is the response

![CleanShot 2024-10-20 at 15 25 27](https://github.com/user-attachments/assets/029b1280-c689-4310-bc64-b8c04e2b4ec8)

3. As user type is doctor, we would be getting doctor_profile_id as 1 and empty doctor profile with id is also created as shown below

irb(main):001> User.all
  User Load (1.4ms)  SELECT "users".* FROM "users" /* loading for pp */ LIMIT $1  [["LIMIT", 11]]
=> [#<User id: 1, email: "dummy1@email.com", created_at: "2024-10-20 09:55:22.010015000 +0000", updated_at: "2024-10-20 09:55:22.010015000 +0000", jti: "f5fcc864-1909-4d91-8bb3-1a454ff1b476", patient_profile_id: nil, doctor_profile_id: 1>]

DoctorProfile.all
  DoctorProfile Load (0.8ms)  SELECT "doctor_profiles".* FROM "doctor_profiles" /* loading for pp */ LIMIT $1  [["LIMIT", 11]]
=> 
[#<DoctorProfile:0x000000010e753a78
  id: 1,
  name: nil,
  gender: nil,
  date_of_birth: nil,
  specialization_id: nil,
  degree_id: nil,
  institute_id: nil,
  year_of_completion: nil,
  year_of_experience: nil,
  created_at: Sun, 20 Oct 2024 09:55:21.850845000 UTC +00:00,
  updated_at: Sun, 20 Oct 2024 09:55:21.850845000 UTC +00:00>]

4. Similarly, if we provide user_type as patient we would get patient_profile_id as 1 and patient db is created with new patient

![CleanShot 2024-10-20 at 15 30 00](https://github.com/user-attachments/assets/67ef37a5-6a5a-4bd1-b780-6689de1a5b70)

5. Verifying the db 

[#<PatientProfile:0x000000010f1327b0
  id: 1,
  name: nil,
  gender: nil,
  date_of_birth: nil,
  blood_group: nil,
  address_id: nil,
  created_at: Sun, 20 Oct 2024 09:59:47.758600000 UTC +00:00,
  updated_at: Sun, 20 Oct 2024 09:59:47.758600000 UTC +00:00>]


 #<User id: 2, email: "dummy2@email.com", created_at: "2024-10-20 09:59:47.885287000 +0000", updated_at: "2024-10-20 09:59:47.885287000 +0000", jti: "747ddb7b-c1f8-4f37-b85a-e3f45f8a3e2d", patient_profile_id: 1, doctor_profile_id: nil>]
